src/docker/composeappengine.cc

```diff
@@ -12,28 +12,111 @@ ComposeAppEngine::ComposeAppEngine(boost::filesystem::path root_dir, std::string
     : root_{std::move(root_dir)},
       compose_{std::move(compose_bin)},
       docker_client_{std::move(docker_client)},
-      registry_client_{std::move(registry_client)} {
-  boost::filesystem::create_directories(root_);
-}
+      registry_client_{std::move(registry_client)} {}
 
 bool ComposeAppEngine::fetch(const App& app) {
-  boost::filesystem::create_directories(appRoot(app));
-  if (download(app)) {
-    LOG_INFO << "Validating compose file";
-    if (cmd_streaming(compose_ + "config", app)) {
-      LOG_INFO << "Pulling containers";
-      return cmd_streaming(compose_ + "pull --no-parallel", app);
+  boost::filesystem::create_directories(appRoot(app) / MetaDir);
+
+  bool result = false;
+
+  try {
+    AppState state(app, appRoot(app), true);
+
+    if (!download(app)) {
+      state.setState(AppState::State::kDownloadFailed);
+      return result;
+    }
+    state.setState(AppState::State::kDownloaded);
+
+    if (!verify(app)) {
+      state.setState(AppState::State::kVerifyFailed);
+      return result;
+    }
+    state.setState(AppState::State::kVerified);
+
+    if (!pullImages(app)) {
+      state.setState(AppState::State::kPullFailed);
+      return result;
+    }
+    state.setState(AppState::State::kPulled);
+
+    result = true;
+  } catch (const std::exception& exc) {
+    LOG_WARNING << "Failed to get/set App state, fallback to forced fetch: " << exc.what();
+
+    if (download(app) && verify(app) && pullImages(app)) {
+      result = true;
     }
   }
-  return false;
+
+  return result;
 }
 
 bool ComposeAppEngine::install(const App& app) {
-  boost::filesystem::ofstream flag_file(appRoot(app) / NeedStartFile);
-  return cmd_streaming(compose_ + "up --remove-orphans --no-start", app);
+  boost::filesystem::create_directories(appRoot(app) / MetaDir);
```

Comment:
can this be called without this existing? If it doesn't exist should we even bother trying to install?

Reply:
Sure, it doesn't make sense in the case of using only from the compose app manager, which might break some tests, but you're right we should return with an error. On the other hand, such case should never happened if used through the compose app manager.